README.md
Terminal
========

[![Build Status on TravisCI](https://secure.travis-ci.org/xp-forge/terminal.png)](https://travis-ci.org/xp-forge/terminal)
[![XP Framework Module](https://raw.githubusercontent.com/xp-framework/web/master/static/xp-framework-badge.png)](https://github.com/xp-framework/core)
[![BSD Licence](https://raw.githubusercontent.com/xp-framework/web/master/static/licence-bsd.png)](https://github.com/xp-framework/core/blob/master/LICENCE.md)
[![Supports PHP 7.0+](https://raw.githubusercontent.com/xp-framework/web/master/static/php-7_0plus.png)](http://php.net/)
[![Latest Stable Version](https://poser.pugx.org/xp-forge/terminal/version.png)](https://packagist.org/packages/xp-forge/terminal)

The terminal library adds support for color to the `util.cmd.Console` class.

![Screenshot](https://raw.githubusercontent.com/xp-forge/terminal/master/terminal-styles.png)

Styles
------
There are five predefined styles:

```php
use util\cmd\Console;

Console::writeLine('This is <primary>primary</>!');
Console::writeLine('This is <success>success</>!');
Console::writeLine('This is <info>info</>!');
Console::writeLine('This is <warning>warning</>!');
Console::writeLine('This is <danger>danger</>!');
```

Colors
------
For direct color control, the colors can be chosen directly by supplying their names. Foreground and background colors are separated by the `@` sign:

```php
use util\cmd\Console;

Console::writeLine('<red>An error occured</>');
Console::writeLine('<white@green>OK: 100 Tests succeeded</>');
```

The colors' names are black, dark-red, dark-green, dark-yellow, dark-blue, dark-magenta, dark-cyan, gray, dark-gray, red, green, yellow, blue, magenta, cyan and white.

Attributes
----------
There are three more attributes. Not all terminals support these, though!

```php
use util\cmd\Console;

Console::writeLine('<underline>http://localhost</>');
Console::writeLine('<bold>Watch out!</>');
Console::writeLine('<italic>- The XP Framework group</>');

// Can be combined, too
Console::writeLine('<dark-blue,underline>http://localhost</>');
```

Positioning
-----------
The `util.cmd.term.Terminal` class allows clearing the screen, positioning the cursor and writing text there.

![Screenshot](https://raw.githubusercontent.com/xp-forge/terminal/master/terminal-clock.png)

```php
use util\cmd\term\Terminal;

$message= 'Hello from the middle of the screen';

// Center text in third line
$dim= Terminal::size();
$x= ($dim[0] - strlen($message)) / 2;
$y= 3;

Terminal::clear();
Terminal::write($x, $y, $message);
```

For the more advanced example seen above in the screenshot, have a look at [this code](https://gist.github.com/thekid/d1a38a0743733d2270968d7eac8f43c9).

ChangeLog.md
Terminal change log
===================

## ?.?.? / ????-??-??

## 1.0.1 / 2020-10-04

* Fixed "Array and string offset access syntax with curly braces is no
  longer supported"
  (@thekid)

## 1.0.0 / 2020-04-04

* Added support for XP 10 - @thekid
* Dropped support for PHP 5, now requires PHP 7.0 mininum! - @thekid

## 0.6.0 / 2018-04-02

* Added compatibility with XP 9 - @thekid
* Fixed compatiblity with PHP 7.2 - @thekid

## 0.5.0 / 2016-09-05

* Changed `Console::writef()` and `writeLinef()` to only tokenize
  format string and not arguments
  (@thekid)
* Changed non-string arguments to `Console::write()` and `writeLine()`
  to no longer be tokenized
  (@thekid)

## 0.4.0 / 2016-09-04

* **Heads up: Dropped PHP 5.5 support - now requires PHP 5.6 minimum!**
  (@thekid)
* Fixed writing multiple arguments with Console. See PR #3 - @thekid
* Added compatibility with XP 8 - @thekid

## 0.3.0 / 2016-07-05

* Fixed issue #2: Terminal confused by debug output - @thekid

## 0.2.0 / 2016-06-20

* Changed preference to bundled `tput` as using Cygwin variant breaks
  ANSI escape sequences inside a Windows shell
  (@thekid)
* Added method to return terminal size as [colums, lines] tuple
  (@thekid)
* Added methods to clear terminal, position cursor and write text at
  given coordinates
  (@thekid)
* Allowed end tags to be spelled-out. The suggested usage is to write
  `<tag>`...`</>`, but `<tag>`...`</tag>` is fine now, too. This can
  be benefitial when editors have in-string autocompletion. Correct
  nesting is *not* verified!
  (@thekid)

## 0.1.0 / 2016-06-19

* Hello World! First release - @thekid

src/test/php/util/cmd/term/unittest/ConsoleTest.class.php
<?php namespace util\cmd\term\unittest;

use io\streams\MemoryOutputStream;
use lang\Value;
use unittest\Test;
use util\Bytes;
use util\cmd\Console;

class ConsoleTest extends \unittest\TestCase {

  /**
   * Assert formatted input matchs expected outcome
   *
   * @param  string $expected
   * @param  function(): void $block
   * @throws unittest.AssertionFailedError
   */
  private function assertWritten($expected, $block) {
    $out= Console::$out->getStream();
    $buffer= new MemoryOutputStream();
    Console::$out->setStream($buffer);

    try {
      $block();
    } finally {
      Console::$out->setStream($out);
    }

    $formatted= $buffer->getBytes();
    if ($expected !== $formatted) {
      $this->fail('equals', new Bytes($formatted), new Bytes($expected));
    }
  }

  /**
   * Returns an object with a toString() output equaling the given value
   *
   * @param  string $value
   * @return lang.Value
   */
  private function value($value) {
    return newinstance(Value::class, [], [
      'toString'  => function() use($value) { return $value; },
      'hashCode'  => function() { return spl_object_hash($this); },
      'compareTo' => function($value) { return $value === $this ? 0 : 1; }
    ]);
  }

  #[Test]
  public function write() {
    $this->assertWritten("\e[31;1mTest\e[22;39m", function() {
      Console::write('<red>Test</>');
    });
  }

  #[Test]
  public function write_multiple_args() {
    $this->assertWritten("\e[31;1mTest\e[22;39m", function() {
      Console::write('<red>', 'Test', '</>');
    });
  }

  #[Test]
  public function writeLine() {
    $this->assertWritten("\e[31;1mTest\e[22;39m\n", function() {
      Console::writeLine('<red>Test</>');
    });
  }

  #[Test]
  public function writeLine_multiple_args() {
    $this->assertWritten("\e[31;1mTest\e[22;39m\n", function() {
      Console::writeLine('<red>', 'Test', '</>');
    });
  }

  #[Test]
  public function write_object() {
    $this->assertWritten("\e[31;1mTest\e[22;39m", function() {
      Console::write('<red>', $this->value('Test'), '</>');
    });
  }

  #[Test]
  public function write_object_with_end_tag() {
    $this->assertWritten("\e[31;1m</>\e[22;39m", function() {
      Console::write('<red>', $this->value('</>'), '</>');
    });
  }

  #[Test]
  public function writeLine_object() {
    $this->assertWritten("\e[31;1mTest\e[22;39m\n", function() {
      Console::writeLine('<red>', $this->value('Test'), '</>');
    });
  }

  #[Test]
  public function writeLine_object_with_end_tag() {
    $this->assertWritten("\e[31;1m</>\e[22;39m\n", function() {
      Console::writeLine('<red>', $this->value('</>'), '</>');
    });
  }

  #[Test]
  public function writef() {
    $this->assertWritten("\e[31;1mTest\e[22;39m", function() {
      Console::writef('<red>%s</>', 'Test');
    });
  }

  #[Test]
  public function writeLinef() {
    $this->assertWritten("\e[31;1mTest\e[22;39m\n", function() {
      Console::writeLinef('<red>%s</>', 'Test');
    });
  }

  #[Test]
  public function writef_argument_with_end_tag() {
    $this->assertWritten("\e[31;1m</>\e[22;39m", function() {
      Console::writef('<red>%s</>', '</>');
    });
  }

  #[Test]
  public function writeLinef_argument_with_end_tag() {
    $this->assertWritten("\e[31;1m</>\e[22;39m\n", function() {
      Console::writeLinef('<red>%s</>', '</>');
    });
  }
}

src/test/php/util/cmd/term/unittest/TerminalTest.class.php
<?php namespace util\cmd\term\unittest;

use unittest\{Test, Values};
use util\Bytes;
use util\cmd\term\Terminal;

class TerminalTest extends \unittest\TestCase {

  /**
   * Assert formatted input matchs expected outcome
   *
   * @param  string $expected
   * @param  string $input
   * @throws unittest.AssertionFailedError
   */
  private function assertFormatted($expected, $input) {
    $stack= [];
    $formatted= Terminal::format($input, $stack);
    if ($expected !== $formatted) {
      $this->fail('equals', new Bytes($formatted), new Bytes($expected));
    }
  }

  #[Test]
  public function format_empty() {
    $this->assertFormatted('', '');
  }

  #[Test]
  public function format_non_empty() {
    $this->assertFormatted('Test', 'Test');
  }

  #[Test]
  public function format_with_color() {
    $this->assertFormatted("\e[31;1mTest\e[22;39m", '<red>Test</>');
  }

  #[Test]
  public function end_tags_may_be_spelled_out() {
    $this->assertFormatted("\e[31;1mTest\e[22;39m", '<red>Test</red>');
  }

  #[Test]
  public function format_with_color_and_underline() {
    $this->assertFormatted("\e[31m\e[4mTest\e[24m\e[39m", '<dark-red,underline>Test</>');
  }

  #[Test]
  public function format_with_color_and_nested_attribute() {
    $this->assertFormatted("\e[31mTest \e[4mLink\e[24m\e[39m", '<dark-red>Test <underline>Link</></>');
  }

  #[Test, Values(['>>> Sent', '<<< Recieved'])]
  public function issue_2($input) {
    $this->assertFormatted($input, $input);
  }

  #[Test]
  public function raw_xml() {
    $this->assertFormatted('<?xml version="1.0" encoding="utf-8"?>', '<><?xml version="1.0" encoding="utf-8"?></>');
  }

  #[Test]
  public function raw_stacktrace() {
    $this->assertFormatted(
      'at <main>::array_pop() [line 76 of Terminal.class.php] Test',
      '<>at <main>::array_pop() [line 76 of Terminal.class.php]</> Test'
    );
  }
}

src/main/php/util/cmd/term/Output.class.php
<?php namespace util\cmd\term;

use io\streams\OutputStream;
use io\streams\OutputStreamWriter;
use util\Objects;

class Output implements OutputStreamWriter {
  public static $direct;
  private $out= null;

  /**
   * Constructor
   *
   * @param  io.streams.OutputStream $out
   */
  public function __construct($out) {
    $this->out= $out;
  }
  
  /**
   * Return underlying output stream
   *
   * @return io.streams.OutputStream
   */
  public function getStream() {
    return $this->out;
  }

  /**
   * Return underlying output stream
   *
   * @param  io.streams.OutputStream $stream
   * @return void
   */
  public function setStream(OutputStream $stream) {
    $this->out= $stream;
  }

  /**
   * Flush output buffer
   *
   * @return void
   */
  public function flush() {
    $this->out->flush();
  }

  /**
   * Print arguments
   *
   * @param  var... $args
   * @return void
   */
  public function write(... $args) {
    $stack= [];
    foreach ($args as $arg) {
      if (is_string($arg)) {
        $this->out->write(Terminal::format($arg, $stack));
      } else {
        $this->out->write(Objects::stringOf($arg));
      }
    }

    while ($end= array_shift($stack)) {
      $this->out->write($end);
    }
  }

  /**
   * Print arguments and append a newline
   *
   * @param  var... $args
   * @return void
   */
  public function writeLine(... $args) {
    $this->write(...$args);
    $this->out->write("\n");
  }

  /**
   * Print a formatted string
   *
   * @param  string $format
   * @param  var... $args
   * @return void
   */
  public function writef($format, ... $args) {
    $stack= [];
    $formatted= Terminal::format($format, $stack);
    while ($end= array_shift($stack)) {
      $formatted.= $end;
    }

    $this->out->write(vsprintf($formatted, $args));
  }

  /**
   * Print a formatted string and append a newline
   *
   * @param  string $format
   * @param  var... $args
   * @return void
   */
  public function writeLinef($format, ... $args) {
    $this->writef($format, ...$args);
    $this->out->write("\n");
  }
}

src/main/php/util/cmd/term/Terminal.class.php
<?php namespace util\cmd\term;

class Terminal {
  public static $theme= [
    'primary'   => ["\e[37;1m", "\e[22;39m"],
    'success'   => ["\e[32;1m", "\e[22;39m"],
    'info'      => ["\e[36;1m", "\e[22;39m"],
    'warning'   => ["\e[33;1m", "\e[22;39m"],
    'danger'    => ["\e[31;1m", "\e[22;39m"],

    'reset'     => ["\e[0m", ""],
    'bold'      => ["\e[1m", "\e[22m"],
    'italic'    => ["\e[3m", "\e[23m"],
    'underline' => ["\e[4m", "\e[24m"],
  ];

  public static $colors= [
    'black'        => ['%0', '%9'],
    'dark-red'     => ['%1', '%9'],
    'dark-green'   => ['%2', '%9'],
    'dark-yellow'  => ['%3', '%9'],
    'dark-blue'    => ['%4', '%9'],
    'dark-magenta' => ['%5', '%9'],
    'dark-cyan'    => ['%6', '%9'],
    'gray'         => ['%7', '%9'],

    'dark-gray'    => ['%0;1', '22;%9'],
    'red'          => ['%1;1', '22;%9'],
    'green'        => ['%2;1', '22;%9'],
    'yellow'       => ['%3;1', '22;%9'],
    'blue'         => ['%4;1', '22;%9'],
    'magenta'      => ['%5;1', '22;%9'],
    'cyan'         => ['%6;1', '22;%9'],
    'white'        => ['%7;1', '22;%9']
  ];

  /**
   * Returns transitions for a given style sequence
   *
   * @param  string $styles
   * @return string[]
   */
  private static function transition($styles) {
    $set= $unset= '';
    foreach (explode(',', $styles) as $style) {
      if (isset(self::$theme[$style])) {
        $set.= self::$theme[$style][0];
        $unset= self::$theme[$style][1].$unset;
      } else {
        sscanf($style, '%[^@]@%s', $fg, $bg);
        if (isset(self::$colors[$fg])) {
          $set.= "\e[".strtr(self::$colors[$fg][0], '%', '3').($bg ? ';'.strtr(self::$colors[$bg][0], '%', '4') : '').'m';
          $unset= "\e[".strtr(self::$colors[$fg][1], '%', '3').($bg ? ';'.strtr(self::$colors[$bg][1], '%', '4') : '').'m'.$unset;
        }
      }
    }
    return [$set, $unset];
  }

  /**
   * Format a string containing `<styles>`...`</>` sequences
   *
   * @param  string $in
   * @param  string[] $stack
   * @return string
   */
  public static function format($in, &$stack) {
    $offset= 0;
    $length= strlen($in);
    $formatted= '';

    do {
      $p= strcspn($in, '<', $offset);
      $formatted.= substr($in, $offset, $p);
      $offset+= $p + 1;
      if ($offset >= $length) break;

      $e= strcspn($in, '>', $offset);
      $token= substr($in, $offset, $e);
      if ('' === $token) {
        $e= strpos($in, '</>', $offset) - $offset;
        $formatted.= substr($in, $offset + 1, $e - 1);
        $e+= 2;
      } else if ('/' === $token[0]) {
        $formatted.= array_pop($stack);
      } else if (strlen($token) !== strspn($token, 'abcdefghijklmnopqrstuvwxyz0123456789-,@')) {
        $formatted.= substr($in, $offset - 1, $e + 1 + 1);
      } else {
        list($set, $unset)= Terminal::transition($token);
        $formatted.= $set;
        $stack[]= $unset;
      }

      $offset+= $e + 1;
    } while ($offset < $length);

    return $formatted;
  }

  /**
   * Returns size
   *
   * @return  int[] columns, rows
   */
  public static function size() {
    static $options= ['bypass_shell' => true, 'suppress_errors' => true];

    if (false !== ($col= getenv('COLUMNS'))) {
      return [(int)$col, (int)getenv('LINES')];
    } else {
      $replacement= dirname(getenv('XP_EXE')).DIRECTORY_SEPARATOR.'tput.exe';
      $command= is_file($replacement) ? $replacement : 'tput';
      $p= proc_open($command.' -S', [['pipe', 'r'], ['pipe', 'w']], $pipes, null, null, $options);
      fputs($pipes[0], "lines\ncols\n");
      fclose($pipes[0]);
      $lines= fgets($pipes[1]);
      $columns= fgets($pipes[1]);
      fclose($pipes[1]);
      if (0 === proc_close($p)) {
        return [(int)$columns, (int)$lines];
      }
    }
    return [80, 24];
  }

  /**
   * Applies style for a given block
   *
   * @param  string $styles
   * @param  function(): void $block
   * @return void
   */
  public static function styled($styles, $block) {
    list($set, $unset)= self::transition($styles);
    Output::$direct->write($set);
    try {
      $block();
    } finally {
      Output::$direct->write($unset);
    }
  }

  /**
   * Clears terminal
   *
   * @param  string $styles Optional
   * @return void
   */
  public static function clear($styles= null) {
    $styles && Output::$direct->write(self::transition($styles)[0]);
    Output::$direct->write("\e[2J");
  }

  /**
   * Positions cursor
   *
   * @param  int $x
   * @param  int $y
   * @return void
   */
  public static function position($x, $y) {
    Output::$direct->write("\e[".(int)$y.';'.(int)$x.'H');
  }

  /**
   * Resets current line
   *
   * @return void
   */
  public static function reset() {
    Output::$direct->write("\r\e[K");
  }

  /**
   * Writes text at a given position
   *
   * @param  int $x
   * @param  int $y
   * @param  string|string[] $arg Single or multiple lines
   * @return void
   */
  public static function write($x, $y, $arg) {
    foreach ((array)$arg as $n => $line) {
      $stack= [];
      $formatted= self::format($line, $stack);
      while ($end= array_shift($stack)) {
        $formatted.= $end;
      }

      Output::$direct->write("\e[".(int)($y + $n).';'.(int)$x.'H'.$formatted);
    }
  }
}